### PR TITLE
8349-support-vscode-settings-schemas

### PR DESCRIPTION
**Pull request by danarad05**
Signed-off-by: Dan Arad <dan.arad@sap.com>

#### What it does
fixes: #8349 #8673 

#### How to test

Open a workspace and check if there are warnings concerning "vscode://schemas/settings/folder" schema.
image

#### Review checklist

- [X] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by danarad05** [comment]
@akosyakov @vince-fugnitto 
cc: @amiramw 
In regards to issue #8349:
1. I have added registration of workspaceSchema and folderSchema uri (user,launch, task exist already). Is that what you meant in 8349?
2. Also there are some other scopes - 'application' | 'machine' | 'window' | 'resource' | 'language-overridable' | 'machine-overridable' which I'm not sure if should be added.
3. Also, Should I update code using `getCombinedSchema()`?
- for instance `generateTree()` in `preference-tree-generator.ts` has entire logic for building pref tree based on combinedSchema. 
- also `updateUnderlyingData` in `preference-tree-provider.ts` is based on combinedSchema for hiding/showing properties for preferences pages.

Thanks

File: packages/core/src/browser/preferences/preference-contribution.ts [hunk 1]
```diff
@@ -187,9 +189,9 @@ export class PreferenceSchemaProvider extends PreferenceProvider {
             const overridden = this.overriddenPreferenceName(preferenceName);
             if (overridden) {
                 delete this.overridePatternProperties.properties[`[${overridden.overrideIdentifier}]`];
-                delete this.combinedSchema.properties[`[${overridden.overrideIdentifier}]`];
+                this.removePropFromSchemas(`[${overridden.overrideIdentifier}]`);
             } else {
-                delete this.combinedSchema.properties[preferenceName];
+                this.removePropFromSchemas(preferenceName);
```
Comment (amiramw) [inline packages/core/src/browser/preferences/preference-contribution.ts:194]: schema is also manipulated in `updateOverridePatternPropertiesKey`. should manipulate the two new fields there as well?

File: packages/preferences/src/browser/preferences-json-schema-contribution.ts [hunk 2]
```diff
@@ -35,18 +39,36 @@ export class PreferencesJsonSchemaContribution implements JsonSchemaContribution
...
+            fileMatch: this.getFileMatch(scopeStr),
             url: uri.toString()
         });
-        this.schemaProvider.onDidPreferenceSchemaChanged(() =>
-            this.inmemoryResources.update(uri, serializeSchema())
-        );
+
+        this.schemaProvider.onDidPreferenceSchemaChanged(() => {
+            for (const prefScope in PreferenceScope) {
+                if (parseInt(prefScope, 10) >= 0) {
```
Comment (amiramw) [inline packages/preferences/src/browser/preferences-json-schema-contribution.ts:62]: same here

**Comment by amiramw** [comment]
> 1. I have added registration of workspaceSchema and folderSchema uri (user,launch, task exist already). Is that what you meant in 8349?

I think it is good enough scope for this PR. Not sure about 8349. There can be future work done on preference UI like vscode that shows all folders in workspace and their specific preferences.

> 2. Also there are some other scopes - 'application' | 'machine' | 'window' | 'resource' | 'language-overridable' | 'machine-overridable' which I'm not sure if should be added.

Maybe le'ts keep it workspace and folder for now.

> How to test
TBD

There should one less warning on problem view about missing schema for folder.

File: packages/core/src/browser/preferences/preference-contribution.ts (hunk 1 above)
Comment (danarad05) [reply packages/core/src/browser/preferences/preference-contribution.ts:194]: `patternProperties` is only used in generateTree() from combinedSchema (which generates the Preference page tree) so I thought it is redundant to update it also for workspaceSchema/folderSchema. 
If you think I should update it anyway - let me know and I'll do it.
image

File: packages/preferences/src/browser/preferences-json-schema-contribution.ts [hunk 3]
```diff
@@ -35,18 +39,36 @@ export class PreferencesJsonSchemaContribution implements JsonSchemaContribution
     protected readonly preferenceConfigurations: PreferenceConfigurations;
 
     registerSchemas(context: JsonSchemaRegisterContext): void {
-        const serializeSchema = () => JSON.stringify(this.schemaProvider.getCombinedSchema());
-        const uri = new URI('vscode://schemas/settings/user');
-        this.inmemoryResources.add(uri, serializeSchema());
-        const baseName = this.preferenceConfigurations.getConfigName() + '.json';
-        const fileMatch = [baseName, UserStorageUri.resolve(baseName).toString()];
+        for (const scope in PreferenceScope) {
+            if (parseInt(scope, 10) >= 0) {
```
[the remark this answers is not in the document]
Comment (danarad05) [reply packages/preferences/src/browser/preferences-json-schema-contribution.ts]: Done

File: packages/preferences/src/browser/preferences-json-schema-contribution.ts (hunk 2 above)
Comment (danarad05) [reply packages/preferences/src/browser/preferences-json-schema-contribution.ts]: Done

**Comment by danarad05** [comment]
@amiramw see my comments - updated. Please review again.
Thanks

**Comment by amiramw** [comment]
@danarad05 i see now that the warning about folder scehma is gone. However there are more warnings now than before for some reason. Please check.
Before:
image
Now:
image

**Comment by danarad05** [comment]
@amiramw 
> @danarad05 i see now that the warning about folder scehma is gone. However there are more warnings now than before for some reason. Please check.

After "vscode://schemas/settings/folder" is now fixed, builtin extension "vscode.json-language-features" is now able to load it's contents and run validation on it.
image

image

image

image

**Review by amiramw — APPROVED** [review]
For the scope of eliminating constant problem about missing schema and for showing extra warnings coming from settings folder schema this PR is good enough and seems to work fine.

There is additional work for embedding it in preference UI etc. (like in vs code). But it can be done later IMO.

**Pull request merged**